### ISSUE TITLE
Revisions Concept Breaks GitOps

### ISSUE THREAD

**Issue by ellis-bigelow** [kubeflow/kfserving#6]
There has been a large discussion over email about this topic. I'm opening this up to the repo so we have history and increase participation.

**Comment by ellis-bigelow** [comment]
Ellis Bigelow:

Hey @Clive Cox, @Rakesh Kelkar, @Dan Sun, @Lun-kai Hsu
I spent some time chatting with the Knative team about this problem. This is indeed something that they're facing and actively exploring fixes to.
In our discussion, we covered three possible solutions:Use multiple resources (effectively configuration + route)Allow users to set revisionName and persist N revisions into the past.Passthrough multiple configs in the service spec.For #3, the Knative team shutdown this approach (for themselves) from a usability perspective. The configs are already complicated enough, and keeping track of two specs is really cumbersome. This is further complicated in the gitops problem because identifying the correct commit to `git revert` relies on knowledge of both the canary revision and the standard.
In that case, you might as well be reverting to a commit where only a single configuration exists (i.e., no canary exists), and we're back where we started with picking a commit that uses @latest.
For #1, this is actually an option if you're using OSS Knative. Their data model uses 4 resources, and users have the freedom to control all of them. The problem still exists for their most common users which are trying to rely on the "Service" resource alone. We share the same problem that it complicates the user experience and defeats one of our primary purposes to be a "simple ml spec"
For #2, I think this is the most promising, and this is the direction that Knative is going as well. This approach allows users to specify an optional revision name, which gives them full control over their gitops flow. They can use the @latest pattern as specified in the doc, or they can have their CI system name the revisions. I think this is something that @Rakesh Kelkar was alluding to as well. 
There is some challenge with garbage collection of old revisions. Knative allows users to specify # of revisions in the past to keep. If we follow the suggested pattern of passing all annotations to sub-resources, users could configure this themselves. I think if we defaulted ourselves to something like 100, we could find a happy medium. This means that users can rely on `git revert HEAD~100`, which should sanely cover most use cases.
What do you think?

**Comment by ellis-bigelow** [comment]
Rakesh Kelkar:
 (todo reformat)

I think the concept of revision is an implementation detail that we need in order to track multiple resources that are created to serve the spec. I feel likewe need a different named identifier that the user must explicitly specify and use to control partial rollouts and rollbacks. This identifier is optional for users who never want to pin and always rollout changes to 100%, but mandatory to while pinning to allow the system and the user to reason about what is going on and clearly identify variants. 
I assert that we must have full desired state in the spec - partial state (where part of it is based on cluster state) is what breaks gitops and makes it impossible to reason about reverts.
Here is a strawman proposal (feel free to shoot down): Lets say that the we allow the user to specify a variant identifier... (distinct from the internal revision we use to track serving resources):
If not specified the variant defaults to - "laika" (first animal in space) - what the user wants to do is in boldNew model first time deployed - variant is named laika, a revision is created to serve at 100%Update model to replace currently serving model - new revision created for variant laika and goes to 100%Try out new model variant deployed as shadow ie 0% - user must pin laika and name the new variant to do this (say named yuri)Increase rollout %age for Yuri to 50- new and old variants must be named. If the spec changes a new revision of yuri will get rolled outRollout to 100%- remove %age and pin Yuri (current revision of Yuri is now LKG)System keeps last X LKG revisions of variants
git revert 5->44 clearly specified 50% Yuri, 50% Laikia - system complies. Since the spec for Yuri in (4) may be subtly different from the spec in (5), a new revision of Yuri may get created. Saved lkg version of Laika will be used
git revert 5->3 or 4->33 clearly specified 0% Yuri so system will revert 100% to lkg revision of Laika. Spec in commit may result in a new revision of Yuri at 0% that matches the commit
git revert to 2spec matches laika lkg, so that goes to 100%
git revert to 1new revision of laika (or old revision if still saved) goes out to 100%
I chose specific names instead of @golden and @latest because:the meaning of latest is ambiguous and changes from 1->5 abovegolden has a qualitative connotation that also changes with timeWhat do you think?

**Comment by ellis-bigelow** [comment]
Clive Cox: (todo reformat)

Thanks  Ellis,
I think I was thinking for GitOps to remove Revision you would supply both resource definitions for Canary and Current in same resource. But that probably does increase complexity as you would need to allow a list of sub-resource definitions. The CRD operator would reconcile this to what is running and only add/remove Revisions as needed.
I agree that maybe the additions of allowing revisionNames and/or showing them in Status gets us most of the way to what we need even if its not GitOps.

Thanks Rakesh,
 As I understand your proposal its adding these custom revisionNames into the spec. I have some questions on whether this solves GitOps RollbacksOn going from 5->3 then I assume its still built on assumption that "laika" can be found in revisions stored in KNative and if found will be used. Doesn't this make a lot of assumptions: like Garbage collections might have been done, (might be ok if Ellis' suggestion that we set stored revisions to 100)
various rollout versions of Laika were done after commit 3 so its not going to be Laika as existed at commit 2?Sorry if I misunderstood.
Thanks,
Clive

**Comment by ellis-bigelow** [comment]
Rakesh Kelkar:

Hey Clive,

Yes you are right it does rely on laika hanging about. This is a fatal flaw... Laika comes crashing down to earth!!

I agree that the only foolproof way is to have both specs specified. That is basically a denormalized version or separate deployment and service specs. Seems better actually since operations will necessarily be atomic?

-r

**Comment by ellis-bigelow** [comment]
Ellis Bigelow:

Hey All,
Clive and Rakesh, thanks for your thoughts! Rakesh, I like how you're thinking, but it does have the flaw you've already identified. This flaw is due to the fact that Laika and Uri both have un-referenceable revisions under the hood. I think this is actually solved by simply naming revisions. If you specify each revision name, you have a handle to it in each of your git commits.

If you think about it, this is really just a special case of your Laika/Uri proposal, except you're only allowed one "revision" for each "variant". If you take it to its logical conclusion, the concept of "variant" dissolves and you're left with just supplying revisionName.

I feel really strongly against providing both canary and primary in the spec, as that might encourage users to do strange traffic splitting configurations that undermine the design of the feature as a "rollout" tool and not an "experimentation" tool. I do want to find an experimentation story, but I don't think this is the right place.

As far as solving gitops for rollouts, I think we should follow Knative's lead here.

**Comment by ellis-bigelow** [comment]
Rakesh Kelkar:

Shouldn't revisions be immutable? can that be enforced if users control revision names and the associated spec?

**Comment by ellis-bigelow** [comment]
Ellis Bigelow:

Revisions are immutable. The admission controller would reject an existing name.

**Comment by ellis-bigelow** [comment]
Clive Cox:

True. But then the users need to ask - ok so I'm on "mymodel24" so need to call it "mymodel25".  I suppose that isn't that bad. Alternatively, you could ensure only 2 revisions are kept in Knative to allow more freedom on reusing names.

Maybe, can we enforce only "latest" is kept after a full rollout? So in that case you can just call your models "green" and "blue"?

**Comment by ellis-bigelow** [comment]
Dan Sun:

Just thinking along with Raresh's idea, I am wondering if knative serving can track all the revisions which made to full rollout and tag those revisions with its own serials of names. In this way it is easier to get semantics like "rollback to LKG", "rollback to LKG~2", canary between "LKG/candidate" or "LKG~2/candidate", it also reduce the operational risks of getting strange traffic split on two arbitrary revisions. 

-Dan

**Comment by ellis-bigelow** [comment]
Ellis Bigelow:

re: Clive's comment,
I think the generation of these names would probably be done with some CI/CD system (or fairing?). Of course, users could manually increment these names, but I imagine there would be a lot of "whoops, I forgot to update the revision name" and the kubectl apply would fail. Maybe that's okay though. The problem with keeping only two is you break the `git revert` for old revisions, as we discussed earlier.

re: Dan, 
I worry a little bit about the complexity of these naming conventions. Maybe I'm not thinking it through, but I'd hate the user to have to think "was it LKG~2 or LKG~3".

**Comment by ellis-bigelow** [comment]
Rakesh Kelkar: (todo format)

So I've had a principle I have been trying to meet to ensure cluster ops is decoupled from model ops:"desired state" is enough to boot strap a new cluster (that starts from zero state). Cluster state created by past operations is not required.

However I have beenstruggling to understand how it is upheld in a gitops context.. consider the following scenario:Commit 1: serve A1Cluster X starts serving A1Commit 2: server A2Cluster X attempts to spin up A2 but A2 pods keep failingCluster X gives up on A2 but keeps serving A1At this point, cluster Ops rolls out a new cluster Y to replace X...Cluster Y syncs the latest commit from GIT - that is A2Cluster Y attempts to serve A2 but also fails to start podsCluster Y is now not serving anything The only way to get around this is for the cluster or some agent that is monitoring the rollout, to trigger a GIT revert when A2 is marked as failed. Is this a reasonable assumption? Does GitOps mandate the presence of this monitoring agent?
Also the x% scheme that we have in the spec right now also breaks this principle (unless as Clive suggests both revision specs are included in the commit). 
Is the stated principle impossible to achieve? Does cluster Y have to sync cluster X LKG state in order to be transparent to users? 
Alternate solves everything proposal (or has another annoying fatal flaw @Clive Cox?)Would it be reasonable instead to just support a spec that has upto two revisions specs slots in it 0, 1 and a knob to switch which one should be primary? Kind of  like how key rotation works? Can also include %age distribution between the two?
In terms of the effective user data model there are two entities, deployment and service... all that we are doing in this spec is denormalizing them into one document to make them "easier" to use... the denormalized view is by-value and we support by-reference in addition we might end up with a nice default but allow those who want separate files with separate edit histories the option to do that too. Strong opinions, lightly held?
-r

**Comment by ellis-bigelow** [comment]
Ellis Bigelow: (todo format)

I think you're right Rakesh, during rollout mode, you lose blind re-instantiation of the resources in another cluster. This is a fundamental issue of the revision pattern. 
I've played a little bit with a "canary" semantic and you'd end up with something like this:apiVersion: "serving.kubeflow.org/v1alpha1"kind: "Service"metadata: name: "myModel" annotations: cloud.google.com/gke-accelerator: nvidia-tesla-t4spec: minReplicas: 1 maxReplicas: 2 tensorflow: modelUri: "gs://mybucket/mymodel" resources: requests: memory: 1Gi cpu: 400m nvidia.com/gpu: 1 rolloutPercent: 10 canary: minReplicas: 1 maxReplicas: 2 tensorflow: modelUri: "gs://mybucket/mymodel" resources: requests: memory: 1Gi cpu: 400m nvidia.com/gpu: 1

It's not great, but it may be out only option if we want to enable the blind re-instantiation pattern you're talking about. I've never seen this pattern in practice, so I wonder how possible it actually is:For things like LoadBalancer Services, IP configurations would be brokenFor things that use ValidatingAdmissionControllers to check pointer in specs of dependent resources, you have an order of operations problem in the apply, which is huge combinatorial explosion which would make this pattern infeasible.From the managed service perspective, it certainly would be nice if I didn't have to think about user's configurations' state during cluster failover. I also like the idea of avoiding the whole "revision name" problem.

**Comment by ellis-bigelow** [comment]
Ellis Bigelow:

Also FYI,
Knative's Service Spec is going to change: https://github.com/knative/serving/milestone/18. We should take a look through as they're likely discussing similar things.

**Comment by ellis-bigelow** [comment]
Rakesh Kelkar: (todo format)

Thanks for the pointer to kn changes, will look.
On the concerns you listed, 
1. Why would it break loadbalancer IP config?
I see the spec as a denormalized version of the underlying resources:Deployment1
Deployment2Service (with deployment selector)
So its just syntactic convenience, because in the minimal case you only specify one spec and get the deployment for free. Also when you update the file, you don't have to remember to delete the old deployment.

2. order of operationsCan you elaborate on the order of operations problem? I agree that the code would be more complex in that it has to deal with a list - but I dont understand the ordering issue..?
-r

**Comment by ellis-bigelow** [comment]
Dan Sun:

Instead of having both revisions on spec, another option maybe is to sync all the spec revision changes to a database(audit trails) and you can recover the LKG state from there in another cluster.
Dan

**Comment by ellis-bigelow** [comment]
Rakesh Kelkar:

My read of your comment is a mechanism to store cluster state and then relying on that state. That is certainly a possibility and would require a new cluster to sync state from existing cluster LKG state before syncing users desired state. Just wondering if that can be reasonably avoided?

**Comment by ellis-bigelow** [comment]
Clive Cox:

I think having a dependency on syncing state from somewhere external adds a lot of external complexity. 
I still like the "gitops" friendly main and canary in same spec...
Echoing Rakesh, can you expand Ellis on the two problems you mentioned?

**Comment by ellis-bigelow** [comment]
Dan Sun: todo format

I like the idea of keeping the LKG on the spec but in practice it does rely on user doing the right thing otherwise the complexity of validation can be a explosion as Ellis said.
For example1 : initially you have M1 and rollout 1002 : canary rollout m2, now you have both M1 M2 on the spec and set percent 20%3 : you decided m2 is not good then you update m2 to m3 4 : m3 looks good then update to 100% and remove m1 have m3 become the LKG
Here is the question what if user did not remove m1 and directly update m3 to m4, we may lose the LGK state ? Just do not know what the validation logic would be here.

**Comment by ellis-bigelow** [comment]
Rakesh Kelkar: todo format

The specs can be validated for syntactical correctness only.  Not sure the spec can be validated for semantic correctness. 
So moving m3 to 100% => making it primary, if the user now updates m3->m4 then the desired state is m4 primary, m1 secondary. If m4 fails to rollout the system will attempt to rollback to m1. This is "correct" per the spec.
The key here is that correctness is maintained and the user is provided the mechanism to specify both lkg and canary. The syntax should make doing the right thing easy, but I don't think we can prevent the user from making a semantic mistake. 

An even worse situation is if the user in your example updates m1 to m3 (when m2 is not good) and then it turns out m3 is not good too - this will put the cluster into a bad state (spec says m2 and m3 both of which are bad) and alerts should fire.
For bad situations, alerts should fire quickly and loudly, and well behaved, responsible CI/CD systems that are monitoring the rollout should revert the commit triggering a roll-forward to LKG state. The cluster may keep state (keep serving m1) but the contract with the user must be that this state cannot be relied upon and a revert is required asap.
I am tempted to suggest that if revisions ids are some kind of spec content hash, we could have an optional check to ensure that whatever is set as secondary exists on the cluster. But that also couples the rollout dependent to cluster state... and would at the very least require a cluster to differentiate steady state from bootstrap. Further this will make things will get even more complex for multiple clusters (with potentially different states) across availability zones or scale units...  

-r
-r

**Comment by ellis-bigelow** [comment]
Ellis Bigelow:

This is a really awesome discussion. I will migrate this to an issue for kfserving and we can continue in there.

**Comment by ellis-bigelow** [comment]
ABOVE COMMENTS ARE FROM EMAIL DISCUSSION 
======================================
BELOW COMMENTS ARE GITHUB DISCUSSION

**Comment by ellis-bigelow** [comment]
Finally off the plane :). A couple of you asked me to follow up on previous points.

I was playing devil's advocate against the idea of whether or not GitOps as we're discussing it is even feasible. Let's take the assumption that we've committed a set of YAML files to a repo and want to apply them to separate clusters.

Problem 1)
For LoadBalancer Services, at least in GKE, you apply a static IP to them. This means that if the configuration is replicated to multiple clusters, one will succeed and one will fail. 

Problem 2)
Many resources reference other resources. Often they do this using ValidatingAdmissionControllers which assert that the resource exists before you point to it. If you're trying to copy a git repo of YAMLs, anything that's asserting existence of other resources will fail. The number of times you have to retry kubectl apply ./**/* to get a successful ordering of these dependencies is combinatorial.

All I'm really trying to say is that there are other common examples of things in the kubernetes universe that make GitOps challenging. That said, I don't think it's a great excuse to be a part of the problem.

**Comment by ellis-bigelow** [comment]
I've compiled some thoughts below. I see saving state to another source of truth as a non-starter. I think there are two remaining options -- please correct me if I've missed or misunderstood one from our discussion!

## Approach A
```
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "KFService"
metadata:
  name: "myModel"
  annotations:
    cloud.google.com/gke-accelerator: nvidia-tesla-t4
spec:
  rolloutPercent: 10
  revisions: ["my-old-rev", "@latest"]
  minReplicas: 1
  maxReplicas: 2
  tensorflow:
    modelUri: "gs://mybucket/mymodel"
    resources:
      requests:
        memory: 1Gi
        cpu: 400m
        nvidia.com/gpu: 1
```

## Approach B
```
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "KFService"
metadata:
  name: "myModel"
  annotations:
    cloud.google.com/gke-accelerator: nvidia-tesla-t4
spec:
  minReplicas: 1
  maxReplicas: 2
  tensorflow:
    modelUri: "gs://mybucket/mymodel"
    resources:
      requests:
        memory: 1Gi
        cpu: 400m
        nvidia.com/gpu: 1
  canary:
    rolloutPercent: 10
    minReplicas: 1
    maxReplicas: 2
    tensorflow:
      modelUri: "gs://mybucket/mymodel-2"
      resources:
        requests:
          memory: 1Gi
          cpu: 400m
          nvidia.com/gpu: 1
```

## TL;DR 
Neither solution is ideal, I'm leaning option B.

## Deep Dive
Firstly, I think the levels of tradeoffs we're making are reasonable given that we're dealing with a reasonably complex feature. The more I think about it, the more I feel that approach A is simpler on the surface, but more complex underneath. As far as approach B goes, I don't like how canary is a sibling of tensorflow. This leads to nasty diffs when moving from canary to primary. I also don't want to nest the "primary" spec under some key so that "canary" can be siblings with the "primary". This will cause the non-canary use cases to be exposed to "primary" key which is very confusing. 

I like that a bunch of concepts go away in approach B, e.g., Revisions, @latest, etc. This also removes weird cases where and old revision has "minNodes" and potentially latent traffic which could cause strange resource usage outside of the users control. I think this simplification might be really nice to users. Another note is that our "status" field suddenly becomes much simpler as we're only dealing with a maximum of 4 (2 x (lastCreated + lastReady)) Revisions under the hood. 

If we follow the pattern of passing through all labels & annotations from the Service to all the subresources, this could potentially cause an issue if someone wants to canary labels/annotations to subresources. This probably isn't a big deal though.

I think Knative is able to get away with their revision approach since users have the ability to modify Revisions or Configurations or Routes if anything goes wrong. I still feel quite strongly about maintaining a single resource semantic, so we may need to make tradeoffs like a bigger spec for the most complex use cases. It's also worth noting that this topic is *still* a huge point of contention among the knative community, so it's reasonable that we make a different decision. We can always pivot if they find a better solution in the meantime. I do realize that we're moving further away from a Knative-like spec as originally proposed, but that's OK!

## Using a Revision Based spec
Pros:
* Supports pinned, canary, and blue-green deployment patterns.
* SimpleSpec. User config will be <20 lines in almost all cases.
* Editing the spec is very clean. User can canary by only editing `revisions` and the field their changing. Clean diffs.

Cons:
* The old Revision's spec is somewhat hidden. Requires past knowledge of configuration.
* Makes use of additional concepts like Revisions and @latest that users must understand.
* Violates git-ops. Cannot migrate arbitrary KFService to another cluster. If deleted, underlying Revisions cannot be recovered.

## Adopting an inline spec 
Pros:
* Supports pinned (rolloutPercent=0), canary, and blue-green (runLatest) deployment patterns.
* Doesn't violate git-ops.  Can't be broken by deleting underlying revisions (or failing over to new cluster)
* Allows complete removal of "Revision" concept, simplifying user experience

Cons:
* User has complete control to violate canary pattern. Can update primary while a canary exists
* Doesn't support n-way revision selection (may be a disguised pro).
* Editing the spec is a little weird, user will need to copy canary spec into primary spec. Unclean diffs.
* Allows users to switch the primary spec while canarying which can be dangerous

## Other thoughts
* In the simple usecase, the spec is identically simple in both approaches (this is good!).
* In the future, could have a point to "KFServiceConfigurations" instead of inline definitions to allow flexibility.

**Comment by rakelkar** [comment]
I dont understand problem 1 you listed above. When you have 1 yaml and 2 clusters, there are two possibilities, depending on how the clusters are related:
a. clusters are independent
b. clusters are related (e.g. setup for across availability zones for HA)

If they are independent, both clusters do their thing and dont care about the other. So in your example both setup an IP.

If they are related, then my assumption is that they have some extra config e.g. scale unit id that allows them to setup the two model instances behind a single higher level construct like an L7 load balancer. On azure this would be Traffic Manager (there are many assumptions in this assertion - however the key point is that cluster admin decides controls the behavior).

For problem 2 - thanks for clarifying. I agree that if there were multiple specs the problem would exist.

**Comment by rakelkar** [comment]
@ellis-bigelow  thanks for the detailed summary and pro/cons. I think the fundamental difference between A & B is that A references revisions that are cluster state while B is fully specified.. I think we can work the spirit of B into different syntax that might alleviate some of the cons you mentioned, specifically unclean diffs - that seems very sad and unavoidable with that particular syntax. The first and last con are anti-patterns that the format allows but would be foolhardy to do (i.e. format should ensure the diff makes this easy to catch in a PR, but if you really want to do it nothing stops you).

Here is an alternative syntax for B that I'd like to throw out...

Notes:
1. the normal spec section if included must be the only spec and is automatically primary and 100%
2. if the normal spec is excluded user can include specs for blue and green. Only one can be primary
3. rolloutPercent can only be set if primary=false
4. rolloutPercent (which is actually trafficPercent IMO) can be 0-99. Cannot be set to 100%. To set to 100%, make primary=true 
4a. rolloutPercent defaults to 0

Scenarios:
1. Simple starter single version - just spec, have fun editing it and rolling out to 100% traffic
2. I need a canary - add blue and green specs. Set primary and on secondary set rollout/traffic=0
3. Hmm, lets try some load - change rollout/traffic=20%
4. Uff, that doesnt work - iterate on secondary spec also adjust rollout if required
4. Looks good ship it! - delete rollout/traffic attribute and set primary to correct values (diff looks great!)


```yaml
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "KFService"
metadata:
  name: "myModel"
  annotations:
    cloud.google.com/gke-accelerator: nvidia-tesla-t4

# NOTE1: can only be set if blue and green specs are omitted. If included - is primary and will be 100%
# spec:
#     minReplicas: 1
#     maxReplicas: 2
#     tensorflow:
#         modelUri: "gs://mybucket/mymodel"
#         resources:
#         requests:
#             memory: 1Gi
#             cpu: 400m
#             nvidia.com/gpu: 1

blue:
    spec:
        # NOTE2: can only be set to true on one of the two!
        isPrimary: true
        minReplicas: 1
        maxReplicas: 2
        tensorflow:
            modelUri: "gs://mybucket/mymodel"
            resources:
            requests:
                memory: 1Gi
                cpu: 400m
                nvidia.com/gpu: 1
green:
    spec:
        isPrimary: false
        # NOTE3: rolloutPercent only be set if isPrimary=false
        # NOTE4: rolloutPercent can only be 0-99 (100 not allowed)
        rolloutPercent: 10
        minReplicas: 1
        maxReplicas: 2
        tensorflow:
        modelUri: "gs://mybucket/mymodel-2"
        resources:
            requests:
            memory: 1Gi
            cpu: 400m
            nvidia.com/gpu: 1
```

**Comment by rakelkar** [comment]
Could also have primary=blue at the top level instead of isPrimary but I like that the isPrimary attribute forces you to go to the spec who's status you are modifying... seems more explicit that way, draws attention even in diffs to exactly what is becoming primary.

**Comment by yuzisun** [comment]
@rakelkar  I think we are getting closer ! This does address the issue for unclear diffs and alleviate some of the potential user errors. Another rule I think we need here is to make the primary spec immutable in blue/green mode, also seems the user specified revision name is needed on spec in the rollback scenario.

```
blue:
    spec:
        # NOTE2: can only be set to true on one of the two!
        isPrimary: true
        userRevisionName: "mymodel1"
        minReplicas: 1
        maxReplicas: 2
        tensorflow:
            modelUri: "gs://mybucket/mymodel"
            resources:
            requests:
                memory: 1Gi
                cpu: 400m
                nvidia.com/gpu: 1
green:
    spec:
        isPrimary: false
        # NOTE3: rolloutPercent only be set if isPrimary=false
        # NOTE4: rolloutPercent can only be 0-99 (100 not allowed)
        userRevisionName: "mymodel2"
        rolloutPercent: 0
        minReplicas: 1
        maxReplicas: 2
        tensorflow:
        modelUri: "gs://mybucket/mymodel-2"
        resources:
            requests:
            memory: 1Gi
            cpu: 400m
            nvidia.com/gpu: 1
```

There is still one case I am struggling to figure out how it would work is that once the `mymodel2` is made primary with full rollout then user finds a regression, in revision approach you can choose to pin to the last good revision and set the rollout percent to 0 but here it seems we are not able to do so after primary is updated. I think we get better user experience with the inline spec but loses some of the flexibility revision approach gives.

**Comment by rakelkar** [comment]
@yuzisun blue green was inspired by a comment you made during the last meeting :)

Why does the user need to specify the revision name? The fact that a revision exists seems like an implementation detail. All that the user asserts is that this spec must be primary. Can the system not match to cluster artifacts (using a primary spec content hash for instance)?

In a way you can consider cluster state as a cache rather than a source of truth. Every rollout is a change (you cant refer to a value in cluster state), but since cluster is smart it will try to use the cache... does that analogy work?

If you accept this reasoning, then the concept of immutable is not valid, since the question immutable with respect to what cannot be answered without referring to cluster state?

Modifications to the primary spec can come from 2 sources - a. make secondary primary, b. change primary in-place. I think you are concerned about the latter. I posit that modifying the primary spec is like saying "I'm confident, do it now"... ofcourse your confidence can be misplaced, however a possible scenario for this is one where I need to do an urgent or safe fix (say increase maxReplicas or resources) on a1 when the state is a1=primary and a2=secondary at 0%. 

The key is that the syntax should make it hard to make this change inadvertently. If you decide to willingly take a risk, well, life will give you feedback if you get too reckless :)

Given this conscious mutability, the solution to your mymodel2 case becomes: if you find a regression after setting a model to primary, you can either modify the primary spec to the LKG spec or revert till you get back to LKG!

**Comment by ellis-bigelow** [comment]
Hey all,
Interesting discussion and a lot to unpack here. It seems like we're converging on something in spirit with Option B. I'm confident we can nail an effective structure for it.

**Firstly**, I think there's a fundamental disagreement on the terminology for blue-green. I understand it as: "A rollout mechanism where a new configuration is deployed to new servers follow by a traffic flip". This is in contrast to an "in-place" deployment where servers are upgraded without changes to routing. In our example, simply updating the normal spec would execute a blue-green deployment. This is basically what k8s deployment gives you out of the box. https://docs.cloudfoundry.org/devguide/deploy-apps/blue-green.html

A canary is a step further than blue-green where instead of a traffic swap, there is a traffic flip. This mechanism isn't automated by default and requires a canary strategy for when to increase the percent. I see this whole discussion as centered around the question: "how to canary".

The three mechanisms can be summarized as:
1. Pinned strategy: rolloutPercent = 0 (rollout resources but not traffic, do manual inspection)
2. Canary strategy: rolloutPercent = [1,99]
3. BlueGreen strategy: rolloutPercent = 100 (can be simplified to a single spec)

By those definitions (i.e., how I understand the industry standard naming), I see your blue/green spec as misnamed.

**Secondly**, I think your spec violates K8s CRD structures. As far as I understand, resources have the 5 standard top level keys shown here: https://kubernetes.io/docs/reference/federation/v1/definitions/#_v1_namespace.

**Thirdly**, I'm not sure I see the value in specifying "isPrimary". Can we not imply "primary" by the top level key? 

Building on your work, what do you think about:

```
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "KFService"
metadata:
  name: "myModel"
spec:
  minReplicas: 1 # can only be set if stable and canary specs are omitted. 
  maxReplicas: 2 # can only be set if stable and canary specs are omitted. 
  tensorflow: # can only be set if stable and canary specs are omitted. 
    modelUri: "gs://mybucket/mymodel"
  stable: # can be used alone or alongside canary
    minReplicas: 1 
    maxReplicas: 2
    tensorflow:
      modelUri: "gs://mybucket/mymodel"
  canary: # must be used alongside stable
    rolloutPercent: 50
    minReplicas: 1 
    maxReplicas: 2
    tensorflow:
      modelUri: "gs://mybucket/mymodel"
```

This should be fairly clean to define using slices.

**Comment by yuzisun** [comment]
Agreed on the naming since we are conflating the two things. `IsPrimary` can be encoded to the type name, but I think it does give some value of cleaner diff and without that you need to copy canary to stable, maybe that's not a big deal. How about the following spec ?
```
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "KFService"
metadata:
  name: "myModel"
spec:
  runlatest:  # can only be set if stable and canary specs are omitted. 
    mlspec: {}
  stable: # can be used alone or alongside canary
    mlspec: {}
  canary: # must be used alongside stable
    rolloutPercent: 50
    mlspec: {}


ModelSpec:
    minReplicas: 1 
    maxReplicas: 2
    tensorflow:
      modelUri: "gs://mybucket/mymodel"
```

**Comment by ellis-bigelow** [comment]
I think that's the closest we've gotten so far.

nit: `runlatest` -> `runLatest`

It's a little "knative-y" since we share that term, but I think that's a good thing. The best part about this is that it results in clean diffs no matter how you edit it.

**Comment by ellis-bigelow** [comment]
Noodling on this further:
```
routeTo:
  <modelSpec>
canary: 
  <modelSpec>
```
"routeTo" makes sense on it's own, but it also makes sense in the context of canarying. Not perfect though...

**Comment by rakelkar** [comment]
Great points @ellis-bigelow thanks for clarifying the terms. Putting both at the same level does improve it. 
@yuzisun I had the isPrimary instead of the type name because that makes it easier to change without inadvertently modifying the spec. However I think the proposed struct can also be edited fairly safely.. 

Here is how I would use it:
1. specify stable
2. specify canary
3. iterate on canary and possibly bump %age
4. delete all the lines from stable modelSpec to canary rolloutPercent and voila canary becomes stable

A few questions and comments:
1. @ellis-bigelow 👍  to merging stable and runLatest. Maybe call it primary or default? (if you feel stable has a qualitative connotation that might be misleading). I don't like routeTo, it has a high handed imperative feel like runLatest (which also feels odd) :)
2. On mature reflection maybe its better to let canary go to 100% (at step 3 above). This additionally allows me to do a single commit that adds a canary at 100% with the confidence that stable will remain if canary rollout fails. With 99% I am forced to do two commits or take a risk. I cant remember why I thought limiting it to 99% was a good idea earlier :-)
3. Can we change rolloutPercent to trafficPercent? I think that makes more sense than rolloutPercent since its traffic that is being controlled. Rollout feels like # of machines/pds but that doesnt make sense in the context of min/max replicas.

**Comment by ellis-bigelow** [comment]
I'm open to rolloutPercent, traffic, trafficPercent, or others. I'm less excited about default primary or default given that they seem very redundant on their own in the non-canary case. That's how I came up with routeTo, as it makes more sense on it's own, similar to runLatest. If we can't find a name that satisfies the intent of both the canary and non-canary use case, we can always keep them as separate "modes" (e.g. runLatest, stable, canary).

Would love to hear from @cliveseldon, or @yuzisun. I'll discuss internally as well. This is always something we can tweak if the structure is known, but it's best to get it right the first try.

**Comment by cliveseldon** [comment]
* I certainly prefer just 2 `modelSpec` locations and not 3 - as 3 is error prone and confusing to me
 * As for `routeTo`, `stable`, `primary` name - I think I prefer `stable` or `primary` as even `routeTo` is confusing with Canary use as the Canary is obviously being `routeTo`-ed!  `stable`/`primary` also implies there can be `canary` in future.
 * I prefer `traffic` or `trafficPercent` if we are enforcing adding up to 100 on both

**Comment by yuzisun** [comment]
- I still prefer to keep 3 because they are two different deployment strategy, we probably do not want to confuse user on the spec and with proper validation it is not an issue. 
- In terms of naming, I feel `stable` and `canary` version are commonly used terminology for canary rollout, or maybe we can call it `canaryStable` and `canaryCandidate` for a better indication that they are logically grouped together? 
- either `rolloutPercent` or `trafficPercent` is fine to me.

re @rakelkar: allowing canary to go 100% does solve my problem in the use case where user may still want to rollback after full rollout, but a bit confusing that stable and canary 100% both exist.

**Comment by ellis-bigelow** [comment]
I will update the PR with latest:

rolloutPercent -> trafficPercent
two subspecs: primary + canary 

We can always rename this, but I want to unblock implementation as I feel we've roughly exhausted this discussion. After some time and introspection, perhaps we'll have new feelings. Sound good?
